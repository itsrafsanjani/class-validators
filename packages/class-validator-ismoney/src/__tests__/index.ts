import 'jest-extended'

import { IsMoney } from '..'
import { validate } from 'class-validator'

type FlakyPriceType =
  | number
  | string
  | {
      amount?: number | string
      currency?: string
    }

describe('IsMoney validation', () => {
  class GuineaPig {
    @IsMoney({ max: 1200 })
    public price?: FlakyPriceType
  }

  class CustomGuineaPig {
    @IsMoney({
      allowNegative: true,
      min: -1100,
      currencies: ['EUR', 'USD', 'CHF'],
    })
    public price?: FlakyPriceType
  }

  class DefaultGuineaPig {
    @IsMoney()
    public price?: FlakyPriceType
  }

  it('accepts valid class', async () => {
    const littlePiggy = new GuineaPig()
    littlePiggy.price = {
      amount: 42,
      currency: 'CHF',
    }
    expect.assertions(1)

    const validationResult = await validate(littlePiggy)
    expect(validationResult).toBeEmpty()
  })

  it('rejects invalid class', async () => {
    const littlePiggy = new GuineaPig()
    littlePiggy.price = {
      amount: 'Enough',
      currency: 'Gold',
    }
    expect.assertions(1)
    const validationResult = await validate(littlePiggy)
    expect(validationResult).toMatchSnapshot()
  })

  it('checks max', async () => {
    const littlePiggy = new GuineaPig()
    littlePiggy.price = {
      amount: 2400,
      currency: 'EUR',
    }
    const errors = await validate(littlePiggy)
    expect(errors).toMatchSnapshot()
  })

  it('rejects negative by default', async () => {
    const littlePiggy = new GuineaPig()
    littlePiggy.price = {
      amount: -2400,
      currency: 'EUR',
    }

    const errors = await validate(littlePiggy)
    expect(errors).toMatchSnapshot()
  })

  it('accepts negative amount when allowNegative is true', async () => {
    const littlePiggy = new CustomGuineaPig()
    littlePiggy.price = {
      amount: -100,
      currency: 'EUR',
    }

    const errors = await validate(littlePiggy)
    expect(errors).toHaveLength(0)
  })

  it('checks min', async () => {
    const littlePiggy = new CustomGuineaPig()
    littlePiggy.price = {
      amount: -2000,
      currency: 'USD',
    }

    const errors = await validate(littlePiggy)
    expect(errors).toMatchSnapshot()
  })

  it('rejects currency which is not in the custom list', async () => {
    const littlePiggy = new CustomGuineaPig()
    littlePiggy.price = {
      amount: 100,
      currency: 'BYN',
    }

    const errors = await validate(littlePiggy)
    expect(errors).toMatchSnapshot()
  })

  it('rejects incorrect values on default settings', async () => {
    const defaultPiggy = new DefaultGuineaPig()
    defaultPiggy.price = {
      amount: -100,
      currency: 'EUR',
    }
    const errors = await validate(defaultPiggy)
    expect(errors).toMatchSnapshot()
  })

  it('accept correct values on default settings', async () => {
    const defaultPiggy = new DefaultGuineaPig()
    defaultPiggy.price = {
      amount: -100,
      currency: 'EUR',
    }
    const errors = await validate(defaultPiggy)
    expect(errors).toMatchSnapshot()
  })

  it('rejects correclty on undefined value', async () => {
    const littlePiggy = new GuineaPig()
    littlePiggy.price = undefined
    const errors = await validate(littlePiggy)
    expect(errors).toMatchSnapshot()
  })
})
