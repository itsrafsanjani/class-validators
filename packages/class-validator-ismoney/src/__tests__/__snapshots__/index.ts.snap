// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsMoney validation accept correct values on default settings 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":-100,\\"currency\\":\\"EUR\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": DefaultGuineaPig {
      "price": Object {
        "amount": -100,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": -100,
      "currency": "EUR",
    },
  },
]
`;

exports[`IsMoney validation checks max 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":2400,\\"currency\\":\\"EUR\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": 2400,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": 2400,
      "currency": "EUR",
    },
  },
]
`;

exports[`IsMoney validation checks min 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":-2000,\\"currency\\":\\"USD\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": CustomGuineaPig {
      "price": Object {
        "amount": -2000,
        "currency": "USD",
      },
    },
    "value": Object {
      "amount": -2000,
      "currency": "USD",
    },
  },
]
`;

exports[`IsMoney validation rejects correclty on undefined value 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "(undefined) is not a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": undefined,
    },
    "value": undefined,
  },
]
`;

exports[`IsMoney validation rejects currency which is not in the custom list 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":100,\\"currency\\":\\"BYN\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": CustomGuineaPig {
      "price": Object {
        "amount": 100,
        "currency": "BYN",
      },
    },
    "value": Object {
      "amount": 100,
      "currency": "BYN",
    },
  },
]
`;

exports[`IsMoney validation rejects incorrect values on default settings 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":-100,\\"currency\\":\\"EUR\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": DefaultGuineaPig {
      "price": Object {
        "amount": -100,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": -100,
      "currency": "EUR",
    },
  },
]
`;

exports[`IsMoney validation rejects invalid class 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":\\"Enough\\",\\"currency\\":\\"Gold\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": "Enough",
        "currency": "Gold",
      },
    },
    "value": Object {
      "amount": "Enough",
      "currency": "Gold",
    },
  },
]
`;

exports[`IsMoney validation rejects negative by default 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsMoney": "({\\"amount\\":-2400,\\"currency\\":\\"EUR\\"}) is not a valid Money object",
    },
    "property": "price",
    "target": GuineaPig {
      "price": Object {
        "amount": -2400,
        "currency": "EUR",
      },
    },
    "value": Object {
      "amount": -2400,
      "currency": "EUR",
    },
  },
]
`;
